File: c/bob/bob.c
#include <stdbool.h>
#include <string.h>
#include <ctype.h>
#include "bob.h"

static bool is_question(char *phrase)
{
	char *end = phrase + strlen(phrase) - 1;

	while (isspace(*end))
		end--;

	return *end == '?';
}

static bool has_letters(char *phrase)
{
	while (*phrase)
	{
		if (isalpha(*phrase))
			return true;
		phrase++;
	}

	return false;
}

static bool is_yelling(char *phrase)
{

	if (!has_letters(phrase))
		return false;

	while (*phrase)
	{
		if (islower(*phrase))
			return false;
		phrase++;
	}

	return true;
}

static bool is_silence(char *phrase)
{

	while (*phrase)
	{
		if (!isspace(*phrase))
			return false;
		phrase++;
	}

	return true;
}

char *hey_bob(char *greeting)
{
	if (is_question(greeting) && is_yelling(greeting))
		return "Calm down, I know what I'm doing!";

	if (is_question(greeting))
		return "Sure.";

	if (is_yelling(greeting))
		return "Whoa, chill out!";

	if (is_silence(greeting))
		return "Fine. Be that way!";

	return "Whatever.";
}

File: python/bob/bob.py
def is_question(text: str) -> bool:
    return text.strip().endswith('?')


def is_yelling(text: str) -> bool:
    return text.isupper()


def is_silence(text: str) -> bool:
    return len(text.strip()) == 0


def response(hey_bob: str) -> str:
    if is_silence(hey_bob):
        return "Fine. Be that way!"
    match (is_yelling(hey_bob), is_question(hey_bob)):
        case (True, True):
            return "Calm down, I know what I'm doing!"
        case (True, False):
            return "Whoa, chill out!"
        case (False, True):
            return "Sure."
    return "Whatever."
